core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -658,14 +599,21 @@ private void install(File snapshotFile) {
       return;
     }
 
+    File snapshotDir = getSnapshotDir(snapshot.getTerm(), snapshot.getIndex());
     long snapshotId = 0L;
     try (Timer.Context ctx = MetricsSystem.timer(MetricKey
-        .MASTER_EMBEDDED_JOURNAL_SNAPSHOT_REPLAY_TIMER.getName()).time();
-         DataInputStream stream =  new DataInputStream(new FileInputStream(snapshotFile))) {
-      snapshotId = stream.readLong();
-      JournalUtils.restoreFromCheckpoint(new CheckpointInputStream(stream), getStateMachines());
+        .MASTER_EMBEDDED_JOURNAL_SNAPSHOT_REPLAY_TIMER.getName()).time()) {
+      SnapshotIdCheckpointed idReader = SnapshotIdCheckpointed.forReading();
+      LOG.debug("Recovered {}", SNAPSHOT_ID);
+      CompletableFuture.allOf(Stream.concat(Stream.of(idReader), getStateMachines().stream())
+          .map(journaled -> journaled.restoreFromCheckpoint(snapshotDir, mJournalPool))
+          .toArray(CompletableFuture[]::new))
+          .join();
+      snapshotId = idReader.getId();
+      mLastSnapshotReplayDurationMs = ctx.stop() / 1_000_000L;
```

Comment:
are you updating the timer twice here? since the try block calls close, and you call stop here